### ISSUE TITLE
Consider using GeneratedDllImport for win32 Interop code

### ISSUE THREAD

**Issue by lindexi** [dotnet/wpf#7040]
Should we update WPF to use GeneratedDllImport where possible?

Reference: https://github.com/dotnet/wpf/issues/7007

See https://github.com/dotnet/runtime/issues/60212

**Comment by singhashish-wpf** [comment]
We should ideally, however before proceeding, we should discuss more on what would be the amount, impact, gain and consequences of this type of a change.

**Comment by RussKie** [comment]
Just be aware that if you decide to use CsWin32 you may have to undo `LibraryImport` (at least until CsWin32 can use that attribute).

**Comment by miloush** [comment]
@lindexi can you summarize what would be the benefits of doing this?

**Comment by ThomasGoulet73** [comment]
I think we should use CsWin32 instead of LibraryImport (New name for GeneratedDllImport) because LibraryImport are still handwritten while CsWin32 generates it for you.

**Comment by lindexi** [comment]
@miloush The CsWin32 can help us to call the system win32 lib. But the wpf should call some wpf native lib. It doesn't look like the two will conflict.

**Comment by McPherran** [comment]
Any existing use of DllImport in select performance-sensitive areas could be updated to LibraryImport. 

- This approach may have less risk even if CsWin32 is updated to match LibraryImport performance without significant impact on API ease of use:
https://github.com/dotnet/wpf/issues/7007#issuecomment-3037038794

- For important performance cases, waiting for CsWin32 to be updated may not be a good idea.
- CsWin32 may not support all cases.
- CsWin32 may become a tooling dependency.
- Use of CsWin32 may raise unnecessary regression testing concerns.

**Comment by rampaa** [comment]
It seems like when using `CsWin32`, setting `allowMarshaling` to `true` will yield user friendly yet potentially less performant API. If it's set to `false`, it will emit API that is as performant as `LibraryImport` but it will not be as user friendly to use. CsWin32 also can't create API with correct nullability annotations, see https://github.com/microsoft/CsWin32/issues/130. So I do agree that using CsWin32 over LibraryImport would be a step down.

**Comment by franchyd** [comment]
> It seems like when using `CsWin32`, setting `allowMarshaling` to `true` will yield user friendly yet potentially less performant API. If it's set to `false`, it will emit API that is as performant as `LibraryImport` but it will not be as user friendly to use. CsWin32 also can't create API with correct nullability annotations, see [microsoft/CsWin32#130](https://github.com/microsoft/CsWin32/issues/130). So I do agree that using CsWin32 over LibraryImport would be a step down.

“Can’t” isn’t quite true.  They just have to annotate *everything* in a single shot because they can’t be inconsistent

**Comment by rampaa** [comment]
> “Can’t” isn’t quite true.

I don't believe It's a useful distinction to make in this case but let me rephrase it for the sake of it: CsWin32 _does not_ create APIs with correct nullability annotations, see https://github.com/microsoft/CsWin32/issues/130.

**Comment by McPherran** [comment]
> ... CsWin32 over LibraryImport would be a step down.

Yes, that may be true at least for some cases and assessing the various cases is itself a task.

If I stay with .NET, I also am probably going to use LibraryImport for now. I say if I stay with .NET: C/C++ doesn't have these issues, but CISA strongly advises using a memory safe language. Some may consider that C++ has complexities, but then we have these complexities simply to call an API. C++20 has modules now. Perhaps WPF could start using modules, as they are an excellent feature. With C# and .NET, we're discussing all these ways to improve performance: interop techniques, stack usage, ref struct, Span<T>, ... C/C++ already has all that automatically, with generally significantly superior performance. WPF uses C++ in some areas. **Perhaps the interop strategy of WPF could be (if it's not already) to use C++ layers that encapsulate Win32/DirectX to provide simple C# interop.** With that approach, CsWin32 APIs for Win32 are largely irrelevant.

I'm open to using CsWin32 if it's improved, and Mr. Arnott has indicated that they would like to improve the usability of the performant cases. The Managed DirectX lib is long deprecated, but in general, I like the idea of the .NET API providing all Win32 APIs including DirectX. This is more reasonable now, I think, because .NET performance has increased beyond what it was a decade ago and AOT is available. Perhaps CsWin32 is part of a potential Win32 .NET API.
